hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
+ *
+ * @param[in,out]   scan_params Pointer to the scanning parameters to be filled.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_get_scan_parameters(hal_ble_scan_params_t* scan_params, void* reserved);
+
+/**
+ * Start scanning nearby BLE devices.
+ *
+ * @param[in] callback  The callback function to handle the scan results.
+ * @param[in] context   The callback function context.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_start_scan(on_ble_scan_result_cb_t callback, void* context, void* reserved);
+
+/**
+ * Check if BLE is scanning nearby devices.
+ *
+ * @returns     true if it is scanning, otherwise false.
+ */
+bool ble_gap_is_scanning(void);
+
+/**
+ * Stop scanning nearby BLE devices.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_stop_scan(void);
+
+/**
+ * Connect to a peer BLE device.
+ *
+ * @param[in]   address Pointer to the peer BLE identity address.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_connect(const hal_ble_addr_t* address, void* reserved);
+
+/**
+ * Check if BLE is connecting with peer device.
+ *
+ * @returns true if connecting, otherwise false.
+ */
+bool ble_gap_is_connecting(void);
+
+/**
+ * Check if BLE is connected with peer device.
+ *
+ * @param[in]   address Pointer to the peer BLE identity address.
+ *                      If NULL, it returns true as long as there has on-going connections.
+ *
+ * @returns true if connected, otherwise false.
+ */
+bool ble_gap_is_connected(const hal_ble_addr_t* address);
+
+/**
+ * Cancel the ongoing procedure that connecting to the peer BLE device.
+ *
+ * @returns true if success, otherwise false.
+ */
+int ble_gap_connect_cancel(void);
+
+/**
+ * Terminate BLE connection.
+ *
+ * @param[in]   conn_handle BLE connection handle. If BLE peripheral role, it is ignored.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_disconnect(uint16_t conn_handle, void* reserved);
+
+/**
+ * Update connection parameters. In the central role this will initiate a Link Layer connection parameter update procedure.
+ * In the peripheral role, this will send the corresponding request and wait for the central to perform the procedure.
+ *
+ * @param[in]   conn_handle BLE connection handle.
+ * @param[in]   conn_params Pointer to a hal_ble_conn_params_t structure with the desired parameters.
+ *                          If NULL is provided on a peripheral role, the parameters in the PPCP characteristic of the GAP
+ *                          service will be used instead. If NULL is provided on a central role and in response to a
+ *                          connection parameters update request, the peripheral request will be rejected.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_update_connection_params(uint16_t conn_handle, const hal_ble_conn_params_t* conn_params, void* reserved);
```

Comment:
Consider removing `reserved` argument as `hal_ble_conn_params_t` is already versioned.